src/js/client/editprefset-viewport.js

```diff
@@ -0,0 +1,681 @@
+            args: ["{that}", "{arguments}.0", "{arguments}.1"] // context, commonTerm
+        },
+        lookupProductPrefValue: {
+            funcName: "gpii.devpmt.lookupProductPrefValue",
+            args: ["{that}", "{arguments}.0", "{arguments}.1", "{arguments}.2"] // context, product, settingTerm
+        },
+        onDeleteContext: {
+            funcName: "gpii.devpmt.onDeleteContext",
+            args: ["{that}", "{arguments}.0"] // event
+        }
+    },
+    listeners: {
+        "onCreate": {
+            funcName: "gpii.devpmt.npsetInit",
+            args: ["{that}"] //"onCreate listener"]
+        },
+        "onMarkupRendered": [
+            {
+                "this": "{that}.dom.topbarSaveButton",
+                "method": "click",
+                args: ["{that}.events.openConfirmSaveDialog.fire"]
+            },
+            {
+                "this": "{that}.dom.deleteContextButtons",
+                "method": "click",
+                args: ["{that}.onDeleteContext"]
+            },
+            {
+                "this": "{that}.dom.devModeIcon",
+                "method": "click",
+                args: ["{that}.toggleDevModeView"]
+            },
+            {
+                funcName: "gpii.devpmt.updateFoundationSticky",
+                args: []
+            },
+            {
+                func: "{that}.initSettingTableWidgets"
+            }
+        ]
+    },
+    modelListeners: {
+        "flatPrefs": [{
+            func: "{that}.updateMetadataFromPrefs",
+            args: ["{that}"]
+        },
+        {
+            func: "{that}.commonTermUsageCounts"
+        },
+        {
+            func: "{that}.reRender",
+            excludeSource: ["init"]
+        }
+        ],
+        "unsavedChangesExist": {
+            func: "{that}.reRender",
+            excludeSource: ["init"]
+        },
+        "npsetApplications": {
+            func: "{that}.reRender",
+            excludeSource: ["init"]
+        }
+    }
+});
+
+/**
+ * Common term filter counts
+ *
+ * Returns the total number of common terms, and then the maximum
+ * number that is set in the NP Set. This is used for rendering
+ * information by the settings filters in the form:
+ *
+ * [[My Settings (3)]] [All Settings (225)]
+ *
+ * This provides the user will contextual knowledge before hiding
+ * or showing all settings. This information could also be used
+ * to perhaps only show the first 10 settings or something, and
+ * then expand the rest when the user decides it is appropriate.
+ *
+ * If the user has multiple contexts, this will return the total
+ * number of unique settings keys in the NP Set across all contexts.
+ *
+ * @return (Object) Object with keys `npset` and `all`, such as
+ *                  {
+ *                      npset: 3,
+ *                      all: 225
+ *                  }
+ */
+gpii.devpmt.commonTermUsageCounts = function (that, commonTermsSorted /*, flatPrefs */) {
```

Comment:
Strange double-effect function - should really either work with the model or be a pure function (preferably the latter)

Reply:
This has been factored out into a calculate function with no side effect, and then another to update the model on changes.